rules/directory/directory.bzl

```diff
@@ -0,0 +1,173 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Skylib module containing rules to create metadata about directories."""
+
+load(
+    ":providers.bzl",
+    "DirectoryInfo",
+    "FileOrDirectoryInfo",
+)
+
+def _directory_impl(ctx):
+    # We rely upon "freezing" a dict by converting it to a struct.
+    # Because there are arbitrary keys based on the file path, this could
+    # potentially overlap with to_json / to_proto.
+    if dir(struct()):
+        fail("directory rule requires --incompatible_struct_has_no_methods")
+
+    source_dir = ctx.file.self.path
+    source_prefix = source_dir + "/"
+
+    # Declare a generated file so that we can get the path to generated files.
+    f = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.write(f, "")
+    generated_dir = f.path.rsplit("/", 1)[0]
+    generated_prefix = generated_dir + "/"
+
+    root_metadata = struct(
+        directories = {},
+        files = [],
+        source_path = source_dir,
+        generated_path = generated_dir,
+        human_readable = "@@%s//%s" % (
+            # repo_name is only available in bazel 7+
+            ctx.label.repo_name if hasattr(ctx.label, "repo_name") else ctx.label.workspace_name,
+            ctx.label.package,
+        ),
+    )
+
+    topological = [root_metadata]
+    for src in ctx.files.srcs:
+        prefix = source_prefix if src.is_source else generated_prefix
```

Comment:
You could get away with `bin_dir`, we're in the process of removing `genfiles_dir` . 

Still using `src.root.path` would have less assumptions in the code. At the moment you're assuming files come from one of 2 roots. Technically a tree artifact may also be a root, at the moment we don't allow them in attributes.

Reply:
I investigated the plausibility of it, and I don't think changing it to `src.root.path` works. What I'm about to say I'm not 100% certain on, so please correct me if I'm wrong:

Consider the following example in blaze:
```
# foo/BUILD
directory(..., srcs = ["source_file", ":genfile", ":binfile"])
```

These three files will have the following respective paths:
```
foo/source_file (root is "")
blaze-out/.../foo/binfile (root is "blaze-out/...")
blaze-genfiles/.../foo/genfile (root is "blaze-genfiles/...")
```

There are now three things we need to do:
1) Check that the file is contained within a the package.
1) Calculate the relative path of the file to the package
1) Generate a `DirectoryInfo` provider with appropriate fields

#### Step 1
My existing solution solves this by:
* Passing in "." (the current directory, which is always a source file) to the directory rule to get the source prefix
* Creating a generated file and getting the directory name of that
* Checking if the source path starts with that prefix

This clearly works for source_file and binfile, but doesn't work for genfile. If I understand correctly, your proposal is to instead use `src.root.path`. However, I don't think this will work because:
* `src.root.path` for `genfile` is `blaze-genfiles/...`
* `src.root.path` for a file in an unrelated repository (eg. `@bar//foo:genfile`, when your directory is in `@foo//foo`) in bazel is `external/bar`

Thus, attempting to strip `src.root.path` will ensure that both these examples result in `foo/genfile`. However, one is valid and the other is invalid.

#### Step 2
This part is easy if you can solve step 1, I don't really care about this.

#### Step 3
Here's where things get hard. We now have 3 files, represented by 3 different directories, and only two fields to store them in (`source_path` and `generated_path`). The only way to fix this would be to change `DirectoryInfo` to have an extra field `genfiles_path`, which seems like a bad idea, given that it's only relevant for blaze, and blaze is getting rid of it soon.

Reply:
I think we have three options here:

## Explicitly store each path seperately
* + Works pretty well in bazel
* - Doesn't work great in blaze due to genfiles
* + Easy for a user to get both the source path and the generated path seperately
* Can use a helper function to get the path for a given directory

## Store them together (your `include_paths` idea)
* + Works well in both bazel and blaze
* - If a user does want specifically the source or generated path, it may be difficult
* Can use a helper function to get the path for a given directory.

## Only allow a single path
With this idea, if a user tries to put source files and generated files in the sources for a directory, we'll error out

* + Works well in both bazel and blaze
* + Easy to get the path
* - Can't store both source and generated files in a directory

Off the top of my head, I can't think of any situation where you'd want both source and generated files at the same time. I also think that if you want to do that, requiring a user to just have two separate rules seems pretty reasonable. Based on that, I think option 3 is the best, since it's the simplest to use. Instead of `include_paths`, we'd only have to store a single path.